RFCs/0002-pluggable-discovery.md

```diff
@@ -0,0 +1,568 @@
+
+#### HELLO command
+
+`HELLO` is the **first command that must be sent** to the discovery to tell the name of the client and the version of the pluggable discovery protocol that the client supports.
+The syntax of the command is:
+
+`HELLO <PROTOCOL_VERSION> "<USER_AGENT>"`
+
+- `<PROTOCOL_VERSION>` is the protocol version (currently `1.0.0`)
+
+- `<USER_AGENT>` is the name and version of the client (double-quotes `"` are not allowed)
+
+some examples:
+
+- `HELLO 1.0.0 "Arduino IDE 1.8.13"`
+
+- `HELLO 1.0.0 "arduino-cli 1.2.3"`
+
+the response to the command is:
+
+```JSON
+{
+  "eventType": "hello",
+  "message": "OK"
+}
+```
+
+#### START command
+
+The `START` command initializes and start the discovery internal subroutines. This command must be called before `LIST` or `START_SYNC`. The response to the start command is:
+
+```JSON
+{
+  "eventType": "start",
+  "message": "OK"
+}
+```
+
+#### STOP command
+
+The `STOP` command stops the discovery internal subroutines and possibly free the internally used resources. This command should be called if the client wants to pause the discovery for a while. The response to the command is:
+
+```JSON
+{
+  "eventType": "stop",
+  "message": "OK"
+}
+```
+
+#### QUIT command
+
+The `QUIT` command terminates the discovery. The response to `QUIT` is:
+
+```JSON
+{
+  "eventType": "quit",
+  "message": "OK"
+}
+```
+
+after this output the discovery exits.
+
+#### LIST command
+
+The `LIST` command executes an enumeration of the ports and returns a list of the available ports at the moment of the call. The format of the response is the following:
+
+```
+{
+  "eventType": "list",
+  "ports": [
+    {
+      "address":       <-- THE ADDRESS OF THE PORT
+      "label":         <-- HOW THE PORT IS DISPLAYED ON THE GUI
+      "protocol":      <-- THE PROTOCOL USED BY THE BOARD
+      "protocolLabel": <-- HOW THE PROTOCOL IS DISPLAYED ON THE GUI
+      "properties": {
+                       <-- A LIST OF PROPERTIES OF THE PORT
+      }
+    }, {
+      ...              <-- OTHER PORTS...
+    }
+  ]
+}
+```
+
+The `ports` field contains a list of the available ports.
+
+Each port has:
+
+- an `address` (for example `/dev/ttyACM0` for serial ports or `192.168.10.100` for network ports)
+
+- a `label` that is the human readable form of the `address` (it may be for example `ttyACM0` or `SSH on 192.168.10.100`)
+
+- `protocol` is the protocol identifier (such as `serial` or `dfu` or `ssh`)
+
+- `protocolLabel` is the `protocol` in human readable form (for example `Serial port` or `DFU USB` or `Network (ssh)`)
+
+- `properties` is a list of key/value pairs that represent information relative to the specific port
+
+To make the above more clear let’s show an example with the `serial_discovery`:
+
+```JSON
+{
+  "eventType": "list",
+  "ports": [
+    {
+      "address": "/dev/ttyACM0",
+      "label": "ttyACM0",
+      "protocol": "serial",
+      "protocolLabel": "Serial Port (USB)",
+      "properties": {
```

Comment:
This should be proposed in the `serial_discovery` project: https://github.com/arduino/serial-discovery
I'll update the tool so it will conform with this RFC once we close it.